lib/broadway/producer.ex

```diff
@@ -171,14 +187,63 @@ defmodule Broadway.Producer do
   end
 
   @impl true
+  def handle_info({__MODULE__, :cancel_consumers}, %{rate_limiting: %{} = rate_limiting} = state) do
+    rate_limiting =
+      if :queue.is_empty(rate_limiting.message_buffer) do
+        cancel_consumers(state)
+        rate_limiting
+      else
+        %{rate_limiting | draining?: true}
```

Comment:
We should set this to true on both clauses. The reason is that we need to check the draining? state to avoid calling handle_demand on `handle_next_demand`.